backend/src/services/auth.rs

```diff
@@ -0,0 +1,349 @@
+
+        auth.login(bob_client.id()).await.map_err(|err| {
+            tracing::error!("{err:?}");
+            StatusCode::UNAUTHORIZED
+        })?;
+        auth.client_store.insert(*bob_client.id(), bob_client);
+        tracing::info!("AUTHORIZATION SUCCESSFUL");
+        tracing::info!("Logged in as {:?}", &auth.user());
+    }
+
+    Ok(res)
+}
+
+#[async_trait]
+pub trait Store<Id, Value> {
+    type Error: std::error::Error + Context;
+
+    /// Load `Value` from abstract store
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if a `Value` cpuldn't be loaded
+    async fn load(&self, user_id: &Id) -> Result<Option<Value>, Self::Error>;
+
+    /// Save `Value` into abstract store
+    /// Returns old data, if there was any
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if a `Value` cpuldn't be loaded
+    async fn save(&mut self, user_id: Id, value: Value) -> Result<Option<Value>, Self::Error>;
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct AuthData<User, UserId> {
+    user: Option<User>,
+    user_id: Option<UserId>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct BobUser {
+    login: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct AuthState<User, Id, UserStore, Client> {
+    user_store: UserStore,
+    client_store: HashMap<Id, Client>,
+    _user: PhantomData<User>,
+}
+
+impl<User, UserId, UserStore, Client> AuthState<User, UserId, UserStore, Client> {
+    pub fn new(user_store: UserStore) -> Self {
+        Self {
+            user_store,
+            _user: PhantomData,
+            client_store: HashMap::new(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct AuthStore<User, Id, Client, SessionStore>
+where
+    SessionStore: Store<Id, User>,
+{
+    session: Session,
+    auth_data: AuthData<User, Id>,
+    user_store: SessionStore,
+    client_store: HashMap<Id, Client>,
+}
+
+impl<User, Id, Client, SessionStore> AuthStore<User, Id, Client, SessionStore>
+where
+    User: Clone + Serialize + for<'a> Deserialize<'a> + Sync + Send,
+    Id: Clone + Serialize + for<'a> Deserialize<'a> + Sync + Send,
+    SessionStore: Store<Id, User> + Sync + Send,
+    Client: Send,
+{
+    async fn login(&mut self, user_id: &Id) -> Result<(), AuthError> {
+        if let Some(user) = self
+            .user_store
+            .load(user_id)
+            .await
+            .change_context(AuthError::LoginError)?
+        {
+            self.auth_data.user = Some(user);
+            self.auth_data.user_id = Some(user_id.clone());
+        }
+
+        self.update_session().change_context(AuthError::LoginError)
+    }
+
+    fn logout(&mut self) -> Result<(), AuthError> {
+        self.auth_data = AuthData {
+            user: None,
+            user_id: None,
+        };
+
+        self.update_session().change_context(AuthError::LogoutError)
+    }
+
+    fn user(&self) -> Option<User> {
+        self.auth_data.user.clone()
```

Comment:
You checked the user on every request. In this case it is bad to `clone` the data. Return reference to user data instead